addons/halo/templates/clusterdefinition.yaml

```diff
@@ -14,8 +14,42 @@ spec:
     port: "$(SVC_PORT_tcp-halo)"
   componentDefs:
     - name: halo
-      workloadType: Stateful  
-      characterType: halo      
+      workloadType: Replication  
+      characterType: halo
+      customLabelSpecs:
+        - key: apps.kubeblocks.halo.patroni/scope
+          value: "$(KB_CLUSTER_NAME)-$(KB_COMP_NAME)-patroni$(KB_CLUSTER_UID_POSTFIX_8)"
+          resources:
+            - gvk: "v1/Pod"
+              selector:
+                app.kubernetes.io/managed-by: kubeblocks
+            - gvk: "apps/v1/StatefulSet"
+              selector:
+                app.kubernetes.io/managed-by: kubeblocks
+      rsmSpec:
+        roles:
+          - name: primary
+            accessMode: ReadWrite
+            canVote: true
+          - name: secondary
+            accessMode: Readonly
+            canVote: true
+        roleProbe:
+          customHandler:
+            - image: {{ .Values.image.registry }}/{{ .Values.image.repository}}:{{.Values.image.tag}}
+              command: 
+                - |
+                  Status=$(curl -s http://localhost:8008) && 
+                  role=$(echo $Status | jq .role  | tr -d '"') &&
+                  if [ "$role" = "master" ]; then echo  "primary"; else echo  "secondary"; fi
+          roleUpdateMechanism: DirectAPIServerEventUpdate
+          initialDelaySeconds: 60
+        memberUpdateStrategy: BestEffortParallel
+      serviceRefDeclarations:
```

Comment:
serviceRef is never used. ETCD address is injected by ENV `etcd-service`.  FYI. You may refer to doc for more info on serviceRefs. https://cn.kubeblocks.io/docs/preview/developer-docs/external-component/reference-external-component